File: hard/127. Word Ladder/127. Word Ladder.py
from typing import List


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        words_graph = set(wordList)  # {word: False for word in wordList}

        if endWord not in words_graph:
            return 0

        res = 1
        queue = [beginWord]
        visited = set([beginWord])
        genes = set()
        for s in wordList:
            for k in s:
                genes.add(k)

        answers = []

        while queue:
            next_queue = []

            for current in queue:
                if current == endWord:
                    answers.append(res)

                for i in range(len(current)):
                    for g in genes:
                        if current[i] == g:
                            continue

                        mutated = current[:i] + g + current[i + 1 :]
                        if mutated in words_graph and mutated not in visited:
                            next_queue.append(mutated)
                            visited.add(mutated)

            queue = next_queue
            res += 1

        return min(answers)


solution = Solution()
print(
    solution.ladderLength(
        beginWord="hit",
        endWord="cog",
        wordList=["hot", "dot", "dog", "lot", "log", "cog"],
    )
)

File: hard/127. Word Ladder/127. Word Ladder.go
package main

import "fmt"

func ladderLength(beginWord string, endWord string, wordList []string) int {
	wordSet := make(map[string]bool)
	for _, w := range wordList {
		wordSet[w] = true
	}

	if _, exists := wordSet[endWord]; !exists {
		return 0
	}

	res := 1
	wordQueue := []string{beginWord}

	for len(wordQueue) > 0 {
		levelSize := len(wordQueue)

		for i := 0; i < levelSize; i++ {
			curr := wordQueue[0]
			if curr == endWord {
				return res
			}

			wordQueue = wordQueue[1:]

			for j := 0; j < len(curr); j++ {
				for c := 'a'; c <= 'z'; c++ {
					if byte(c) == curr[j] {
						continue
					}

					mutated := curr[:j] + string(c) + curr[j+1:]
					if _, exists := wordSet[mutated]; exists {
						wordQueue = append(wordQueue, mutated)
						delete(wordSet, mutated)
					}

				}
			}
		}

		res += 1
	}

	return 0
}

func main() {
	fmt.Println(ladderLength("hit", "cog", []string{"hot", "dot", "dog", "lot", "log", "cog"}))
}
